### PR TITLE
livecheck: allow custom url in extract_plist strategy

### PR DESCRIPTION
**Pull request by bevanjkay**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This allows a custom url to be passed to livecheck when using the `:extract_plist` strategy. 

This is useful in edge cases that meet a couple of criteria:
- There is no version information available utilising the other livecheck methods
- There is an unversioned download link that points to the latest version
- There is a versioned download link available that we can use for download

At the moment in these cases we are limited to using the unversioned url OR using the versioned url and skipping the livecheck.

An example of a cask that is currently not functioning as intended is the [`speedify` cask](https://github.com/Homebrew/homebrew-cask/blob/43eadfc/Casks/speedify.rb).

### CODE REVIEW

**Comment by bevanjkay** [comment]
I'm getting a typecheck error, I'd appreciate some direction on how to get it sorted.

```
brew typecheck   
livecheck/strategy/extract_plist.rb:104: Method url does not exist on T.class_of(Homebrew::Livecheck::Strategy::ExtractPlist) https://srb.help/7003
     104 |              url url.to_s
                        ^^^
```

**Review by samford — COMMENTED** [review]
In general, the idea behind this PR makes sense to me. Checking a versioned `url` with `ExtractPlist` would be a waste of effort (as it would simply give the current `version`) and, as you mentioned, it's necessary to check an unversioned `url` in cases like `speedify` to identify the latest version 👍

-----

> I'm getting a typecheck error, I'd appreciate some direction on how to get it sorted.
> 
> ```
> brew typecheck   
> livecheck/strategy/extract_plist.rb:104: Method url does not exist on T.class_of(Homebrew::Livecheck::Strategy::ExtractPlist) https://srb.help/7003
>      104 |              url url.to_s
>                         ^^^
> ```

I tinkered with this and the only way I found to avoid the typecheck error was to use `Cask::CaskLoader.load` with a string, like we do in tests (e.g., [`livecheck/livecheck_spec.rb`](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/livecheck/livecheck_spec.rb#L29-L45)). For example, the existing code would be like this:

```ruby
Cask::CaskLoader.load(+<<-RUBY, config: cask.config)
  cask "livecheck-url-cask" do
    url "#{url}"
  end
RUBY
```

That said, we can simplify this and avoid the `UnversionedCaskChecker` changes if we can create a copy of the cask with a substitute `url` (instead of passing in a supplementary `livecheck_url` cask to use for `UnversionedCaskChecker#installer`). It would be ideal if we could simply replace the underlying `cask.url` `URL::DSL` object but `Cask::DSL#set_unique_stanza` won't allow this (e.g., `cask.url { url }` produces a `Cask::CaskInvalidError (Cask 'speedify' definition is invalid: 'url' stanza may only appear once.)` error).

However, if there aren't any objections to using `#instance_variable_set` in this context, we can sidestep the issue by directly replacing the `@url` instance variable (in the `Cask::DSL` object). I've added a suggestion below that demonstrates this approach but it shouldn't be committed unless/until the approach is approved.

File: Library/Homebrew/livecheck/strategy/extract_plist.rb [hunk 1]
```diff
@@ -85,20 +85,30 @@ def self.versions_from_items(items, regex = nil, &block)
...
-          unversioned_cask_checker = UnversionedCaskChecker.new(cask)
+          if url && url != cask.url.to_s
+            cask_object_for_livecheck = Cask::Cask.new("livecheck-cask", config: cask.config) do
+              url url.to_s
+            end
+
+            unversioned_cask_checker = UnversionedCaskChecker.new(cask, livecheck_url: cask_object_for_livecheck)
+          else
+            unversioned_cask_checker = UnversionedCaskChecker.new(cask)
+          end
```
[the remark this answers is not in the document]
Comment (bevanjkay) [reply Library/Homebrew/livecheck/strategy/extract_plist.rb:110]: @samford Thanks. That is my preferred approach, I just couldn't work out how to change the URL in the Cask object. 😀

**Comment by samford** [comment]
In the interest of merging this, I rebased the branch and pushed a couple commits. The first commit adds a missing documentation comment and adds `url` to the `#find_versions` return hash (to align it with other strategies). The second commit reworks how we handle the `url` argument (and reverts the `UnversionedCaskChecker` changes) per the previous suggestion.

File: Library/Homebrew/livecheck/strategy/extract_plist.rb [hunk 2]
```diff
@@ -82,25 +82,37 @@ def self.versions_from_items(items, regex = nil, &block)
...
           raise ArgumentError, "The #{T.must(name).demodulize} strategy only supports casks." unless T.unsafe(cask)
 
-          match_data = { matches: {}, regex: regex }
+          match_data = { matches: {}, regex: regex, url: url }
+
+          unversioned_cask_checker = if url.present? && url != cask.url.to_s
+            # Create a copy of the `cask` where the `url` is replaced with
+            # the `url` argument
+            cask_copy = Cask::CaskLoader.load(cask.full_name)
+            cask_copy.instance_variable_get(:@dsl).instance_variable_set(:@url, URL::DSL.new(url))
```
Comment (samford) [inline Library/Homebrew/livecheck/strategy/extract_plist.rb:110]: @MikeMcQuaid Any objections to using `#instance_variable_set` here to replace the `@url` instance variable in a copy of the cask? As mentioned above, something like `cask_copy { url }` produces a `Cask::CaskInvalidError`, as `Cask::DSL#set_unique_stanza` prevents `@url` from being set more than once.

Comment (bevanjkay) [reply Library/Homebrew/livecheck/strategy/extract_plist.rb:110]: Thanks @samford

**Review by MikeMcQuaid — COMMENTED** [review]
Need new API for this.

File: Library/Homebrew/livecheck/strategy/extract_plist.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy/extract_plist.rb:110]: > If it matters, it's basically just replicating [what happens in Cask::DSL#set_unique_stanza](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/cask/dsl.rb#L112-L124) but we're outside of the Cask::DSL object, so we also have to use instance_variable_get(:@dsl).

Yeh, I think it's different when it's happening within the class itself.

> Past that, I'm not sure how we would approach this within Cask, other than removing the "stanza may only appear once" restrictions (i.e., bringing it in line with the Formula DSL behavior). Was there anything in particular you were thinking of?

some options:
- allow optionally disabling the "stanza may only appear once" in this case (but not in the cask itself)
- add a new internal API for this case

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy/extract_plist.rb:110]: @samford Yeh, this would work for me!

Comment (MikeMcQuaid) [reply Library/Homebrew/livecheck/strategy/extract_plist.rb:110]: @samford maybe `allow_reassignment: false` by default or something?

**Comment by samford** [comment]
I've rebased this and incorporated the previously-discussed `allow_reassignment` approach in `Cask`, so it's ready for review.

Past that, I added a commit to ensure that livecheck only requires a URL when a `livecheck` block specifies a strategy and the strategy's `#find_versions` method has a required `url` parameter. Up to this point, we've had to add a `url` to `livecheck` blocks using `strategy :extract_plist`, even though `ExtractPlist#find_versions` checks a cask instead of a URL. Now that `ExtractPlist` supports an optional URL, it would be clearer to only include a URL in related `livecheck` blocks when it's intended to override the cask `url`.

**Review by MikeMcQuaid — APPROVED** [review]
Looks great, nice work @samford!

**Comment by samford** [comment]
Thanks, everyone!
